### ISSUE TITLE
Pushing block to fork database that failed to link causes internal "head_block" number increasing, which prevents a correct block being pushed into chain later.

### ISSUE THREAD

**Issue by abitmore** [cryptonomex/graphene#264]
Assert failure in default.log:

```
2015-08-23T05:09:24 th_a:invoke handle_block         handle_block ] Got block #194978 from network                      application.cp
p:348
2015-08-23T05:09:25 th_a:invoke handle_block         handle_block ] Got block #194979 from network                      application.cp
p:348
2015-08-23T05:09:26 th_a:invoke handle_block         handle_block ] Got block #194980 from network                      application.cp
p:348
2015-08-23T05:09:28 th_a:invoke handle_block         handle_block ] Got block #194982 from network                      application.cp
p:348
2015-08-23T05:09:28 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:29 th_a:invoke handle_block         handle_block ] Got block #194983 from network                      application.cp
p:348
2015-08-23T05:09:29 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:30 th_a:invoke handle_block         handle_block ] Got block #194984 from network                      application.cp
p:348
2015-08-23T05:09:30 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:31 th_a:invoke handle_block         handle_block ] Got block #194985 from network                      application.cp
p:348
2015-08-23T05:09:31 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:32 th_a:invoke handle_block         handle_block ] Got block #194986 from network                      application.cp
p:348
2015-08-23T05:09:32 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:33 th_a:invoke handle_block         handle_block ] Got block #194987 from network                      application.cpp:348
2015-08-23T05:09:33 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:34 th_a:invoke handle_block         handle_block ] Got block #194988 from network                      application.cpp:348
2015-08-23T05:09:34 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:35 th_a:invoke handle_block         handle_block ] Got block #194989 from network                      application.cpp:348
2015-08-23T05:09:35 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:37 th_a:invoke handle_block         handle_block ] Got block #194990 from network                      application.cpp:348
2015-08-23T05:09:37 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:38 th_a:invoke handle_block         handle_block ] Got block #194991 from network                      application.cpp:348
2015-08-23T05:09:38 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-08-23T05:09:38 th_a:invoke handle_block         handle_block ] Got block #194981 from network                      application.cpp:348
2015-08-23T05:09:38 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
second_branch_itr != _index.get<block_id>().end():
    {}
    th_a  fork_database.cpp:186 fetch_branch_from

    {"first":"0002f9afa8129d6e7bdf1e6e6ca37026bb3a7169","second":"0002f9a45a96baafbbca197a63d5de6c5352aa17"}
    th_a  fork_database.cpp:217 fetch_branch_from

    {"new_block":{"previous":"0002f9a45a96baafbbca197a63d5de6c5352aa17","timestamp":"2015-08-23T05:09:27","witness":"1.6.13","next_secret_hash":"9952698168381aaa0a90afda5e717d9660c0abbe","previous_secret":"e8c75daf30b792f467adb75a5ec8bc5cdc57e562","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"2058c3d58241c01501a0f07a66b0664194086288012835862897fa862b4d62a7591794d02c5595964000669e87b92b60caffaa908380cd668a18658adb200d5974","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:370
2015-08-23T05:09:39 th_a:invoke handle_block         handle_block ] Got block #194992 from network                      application.cpp:348
2015-08-23T05:09:39 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
second_branch_itr != _index.get<block_id>().end():
    {}
    th_a  fork_database.cpp:186 fetch_branch_from

    {"first":"0002f9b075ed96b2bd84d96de2d4c7c0688fce88","second":"0002f9a45a96baafbbca197a63d5de6c5352aa17"}
    th_a  fork_database.cpp:217 fetch_branch_from

    {"new_block":{"previous":"0002f9afa8129d6e7bdf1e6e6ca37026bb3a7169","timestamp":"2015-08-23T05:09:39","witness":"1.6.47","next_secret_hash":"4ce8a96a38e91893a142b27b7a4fdd6e868a82f0","previous_secret":"bcf6ae7c1545e35eea879dc8ca1a3335efa7143c","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"203ebf7f37198b6a1b99744991a4d76b780257950a988e6000b25cdfacd60b0dbc3915128665ffe0fb9278c2eb7c94e2fe9fd7371e3be878b21bfaf82ee46400ce","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:370
```

At the time,

```
locked >>> info
info
{
  "head_block_num": 194980,
  "head_block_id": "0002f9a45a96baafbbca197a63d5de6c5352aa17",
  "head_block_age": "8 hours old",
  "next_maintenance_time": "8 hours ago",
  "chain_id": "d011922587473757011118587f93afcc314fbaea094fc1055574721b27975083",
  ...
  "entropy": "2dea1cf6c29ee6985f7e4e92eba0ad1d9aa2db3c"
}
locked >>> get_block 194980
get_block 194980
{
  "previous": "0002f9a37c9e78e28381eb6d43cd55a6bc6b4171",
  "timestamp": "2015-08-23T05:09:26",
  "witness": "1.6.56",
  "next_secret_hash": "4ce8a96a38e91893a142b27b7a4fdd6e868a82f0",
  "previous_secret": "bcf6ae7c1545e35eea879dc8ca1a3335efa7143c",
  "transaction_merkle_root": "0000000000000000000000000000000000000000",
  "extensions": [],
  "witness_signature": "204ed505d4b1f4be49e91109e2dc39d507b6728f075393f65aa570930c6aac5c533b686565415e008092d40ba7ca9e1622c2bf0ebf9f4f5c1f81a1b023ebe3a8b3",
  "transactions": []
}
locked >>> get_block 194979
get_block 194979
{
  "previous": "0002f9a2fa5ecda920023b3bb7d57dbfaebde346",
  "timestamp": "2015-08-23T05:09:25",
  "witness": "1.6.85",
  "next_secret_hash": "4ce8a96a38e91893a142b27b7a4fdd6e868a82f0",
  "previous_secret": "bcf6ae7c1545e35eea879dc8ca1a3335efa7143c",
  "transaction_merkle_root": "0000000000000000000000000000000000000000",
  "extensions": [],
  "witness_signature": "202ecfbe0beaaef3baf72eb9b7abbf5a232d0a277aee0c5ba7529ae610cc2a1b3940bc4d905f8c36b3e0454e43d7181a20fe2a18da0acf4f3a07a077640c4ab9cd",
  "transactions": []
}

```

**Comment by abitmore** [comment]
Related p2p.log

```
2015-08-23T05:09:27 p2p:message read_loop on_item_ids_inventor ] received inventory of 1 items from peer 45.55.6.216:1776           node.cpp:2613
2015-08-23T05:09:27 p2p:message read_loop on_item_ids_inventor ] adding item 300da4e06ebbb4fc0845bbd68c6df6474197a408 from inventory message to our list of items to fetch          node.cpp:2647
2015-08-23T05:09:27  p2p:fetch_items_loop     fetch_items_loop ] beginning an iteration of fetch items (7 items to fetch)           node.cpp:1104
2015-08-23T05:09:27  p2p:fetch_items_loop     fetch_items_loop ] requesting item 300da4e06ebbb4fc0845bbd68c6df6474197a408 from peer 45.55.6.216:1776            node.cpp:1123
...
2015-08-23T05:09:37 p2p:terminate_inactive_connections_loop terminate_inactive_c ] Disconnecting peer 45.55.6.216:1776 because they didn't respond to my request for item 300da4e06ebbb4fc0845bbd68c6df6474197a408          node.cpp:1325
2015-08-23T05:09:37 p2p:message read_loop            read_loop ] disconnected 0 exception: unspecified
Operation canceled 
    {"message":"Operation canceled"}
    asio  asio.cpp:37 operator()

    {"len":16}
    p2p  stcp_socket.cpp:111 readsome           message_oriented_connection.cpp:199
...
2015-08-23T05:09:38  p2p:fetch_items_loop     fetch_items_loop ] requesting item 300da4e06ebbb4fc0845bbd68c6df6474197a408 from peer 216.252.204.69:65026            node.cpp:1123
...
2015-08-23T05:09:38 p2p:message read_loop           on_message ] handling message block_message_type 300da4e06ebbb4fc0845bbd68c6df6474197a408 size 172 from peer 216.252.204.69:65026           node.cpp:1651
2015-08-23T05:09:38 p2p:message read_loop process_block_during ] received a block from peer 216.252.204.69:65026, passing it to client          node.cpp:3087
2015-08-23T05:09:38 p2p:message read_loop process_block_during ] Failed to push block 194981 (id:0002f9a520062403bc05eddfe0fc00c11f487692), client rejected block sent by peer          node.cpp:3198
```

**Comment by abitmore** [comment]
And later, in default.log:

```
2015-08-23T05:10:00 th_a:invoke handle_block         handle_block ] Got block #195012 from network                      application.cp
p:348
2015-08-23T05:10:00 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
second_branch_itr != _index.get<block_id>().end():
    {}
    th_a  fork_database.cpp:186 fetch_branch_from

    {"first":"0002f9c4378feed0cd1eb0662b169a43a02d311b","second":"0002f9a45a96baafbbca197a63d5de6c5352aa17"}
    th_a  fork_database.cpp:217 fetch_branch_from

    {"new_block":{"previous":"0002f9c3a3deacb2151e25441e939f94bc8aa809","timestamp":"2015-08-23T05:10:00","witness":"1.6.33","next_sec
ret_hash":"15a12f3f7d6391020c29cca55b07bca920721d3e","previous_secret":"d360d760cacf4ce1e1d5ca8b272f4ea3ecf71d27","transaction_merkle_
root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"207382db61d509a3d932b3a967715c435850761a2e78785a
6a4f06f84d8cbf43826283a4ca1b45ce4d9955b65f7eb88228764d54ffad7cb88567fb6f2d6b1f03ab","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:370
2015-08-23T05:10:00 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
item->num > std::max<int64_t>( 0, int64_t(_head->num) - (_max_size) ): attempting to push a block that is too old
    {"item->num":194981,"head":195012,"max_size":11}
    th_a  fork_database.cpp:69 _push_block

    {"new_block":{"previous":"0002f9a45a96baafbbca197a63d5de6c5352aa17","timestamp":"2015-08-23T05:09:27","witness":"1.6.13","next_secret_hash":"9952698168381aaa0a90afda5e717d9660c0abbe","previous_secret":"e8c75daf30b792f467adb75a5ec8bc5cdc57e562","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"2058c3d58241c01501a0f07a66b0664194086288012835862897fa862b4d62a7591794d02c5595964000669e87b92b60caffaa908380cd668a18658adb200d5974","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:370
2015-08-23T05:10:00 th_a:invoke handle_block         handle_block ] Got block #195012 from network                      application.cpp:348
2015-08-23T05:10:00 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
second_branch_itr != _index.get<block_id>().end():
```

in p2p.log:

```
2015-08-23T05:10:00  p2p:fetch_items_loop     fetch_items_loop ] beginning an iteration of fetch items (6 items to fetch)           node.cpp:1104
2015-08-23T05:10:00 p2p:message read_loop           on_message ] handling message block_message_type 300da4e06ebbb4fc0845bbd68c6df6474197a408 size 172 from peer 176.9.234.167:57727            node.cpp:1651
2015-08-23T05:10:00 p2p:message read_loop process_block_during ] received a sync block from peer 176.9.234.167:57727            node.cpp:3073
2015-08-23T05:10:00 p2p:process_backlog_of_sync_blocks process_backlog_of_s ] in process_backlog_of_sync_blocks         node.cpp:2959
2015-08-23T05:10:00 p2p:process_backlog_of_sync_blocks process_backlog_of_s ] currently 0 blocks in the process of being handled            node.cpp:2965
2015-08-23T05:10:00 p2p:process_backlog_of_sync_blocks process_backlog_of_s ] currently 9 sync items to consider            node.cpp:2995
2015-08-23T05:10:00 p2p:process_backlog_of_sync_blocks process_backlog_of_s ] currently 8 sync items to consider            node.cpp:2995
2015-08-23T05:10:00 p2p:process_backlog_of_sync_blocks process_backlog_of_s ] leaving process_backlog_of_sync_blocks, 1 processed           node.cpp:3056
2015-08-23T05:10:00 p2p:process_backlog_of_sync_blocks trigger_fetch_sync_i ] Triggering fetch sync items loop now          node.cpp:1081
2015-08-23T05:10:00 p2p:fetch_sync_items_loop fetch_sync_items_loo ] beginning another iteration of the sync items loop         node.cpp:1020
2015-08-23T05:10:00 p2p:fetch_sync_items_loop fetch_sync_items_loo ] no sync items to fetch right now, going to sleep           node.cpp:1070
2015-08-23T05:10:00 p2p:send_sync_block_to_node_delegate send_sync_block_to_n ] in send_sync_block_to_node_delegate()           node.cpp:2777
2015-08-23T05:10:00 p2p:send_sync_block_to_node_delegate send_sync_block_to_n ] Failed to push sync block 194981 (id:0002f9a520062403bc05eddfe0fc00c11f487692): client rejected sync block sent by peer: {"code":10,"name":"assert_exception","message":"Assert Exception","stack":[{"context":{"level":"error","file":"fork_database.cpp","line":69,"method":"_push_block","hostname":"","thread_name":"th_a","timestamp":"2015-08-23T05:10:00"},"format":"item->num > std::max<int64_t>( 0, int64_t(_head->num) - (_max_size) ): attempting to push a block that is too old","data":{"item->num":194981,"head":195012,"max_size":11}},{"context":{"level":"warn","file":"db_block.cpp","line":176,"method":"_push_block","hostname":"","thread_name":"th_a","timestamp":"2015-08-23T05:10:00"},"format":"","data":{"new_block":{"previous":"0002f9a45a96baafbbca197a63d5de6c5352aa17","timestamp":"2015-08-23T05:09:27","witness":"1.6.13","next_secret_hash":"9952698168381aaa0a90afda5e717d9660c0abbe","previous_secret":"e8c75daf30b792f467adb75a5ec8bc5cdc57e562","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"2058c3d58241c01501a0f07a66b0664194086288012835862897fa862b4d62a7591794d02c5595964000669e87b92b60caffaa908380cd668a18658adb200d5974","transactions":[]}}},{"context":{"level":"warn","file":"application.cpp","line":379,"method":"handle_block","hostname":"","thread_name":"th_a","timestamp":"2015-08-23T05:10:00"},"format":"","data":{"blk_msg":{"block":{"previous":"0002f9a45a96baafbbca197a63d5de6c5352aa17","timestamp":"2015-08-23T05:09:27","witness":"1.6.13","next_secret_hash":"9952698168381aaa0a90afda5e717d9660c0abbe","previous_secret":"e8c75daf30b792f467adb75a5ec8bc5cdc57e562","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"2058c3d58241c01501a0f07a66b0664194086288012835862897fa862b4d62a7591794d02c5595964000669e87b92b60caffaa908380cd668a18658adb200d5974","transactions":[]},"block_id":"0002f9a520062403bc05eddfe0fc00c11f487692"},"sync_mode":true}}]}          node.cpp:2813
2015-08-23T05:10:00 p2p:send_sync_block_to_node_delegate send_sync_block_to_n ] disconnecting client 176.9.234.167:57727 because it offered us the rejected block           node.cpp:2928
```

**Comment by abitmore** [comment]
The second trial to push block 194981 failed because `"head":195012`, it's wrong because that 195012 got rejected by the client a second ago.

**Comment by abitmore** [comment]
Title updated to make the issue clearer.
Pushing block to fork database that failed to link causes internal "head_block" number increasing, which prevents a correct block being pushed into chain later.

**Comment by abitmore** [comment]
Still have same problem with `test2` tag, failed to push block 10046:

```
2015-09-01T11:21:55 th_a:invoke handle_block         handle_block ] Got block #10045 from network                       application.cp
p:356
2015-09-01T11:21:55 th_a:invoke handle_block   on_objects_changed ] ids: ["2.8.10045","1.6.92","1.13.88","2.1.0"]
        api.cpp:867
2015-09-01T11:21:55 th_a:invoke handle_block   on_objects_changed ] ids: ["2.8.10045","1.6.92","1.13.88","2.1.0"]
        api.cpp:867
2015-09-01T11:22:05 th_a:invoke handle_block         handle_block ] Got block #10047 from network                       application.cp
p:356
2015-09-01T11:22:05 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-09-01T11:22:10 th_a:invoke handle_block         handle_block ] Got block #10048 from network                       application.cp
p:356
2015-09-01T11:22:10 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
...
...
...
2015-09-01T11:24:30 th_a:invoke handle_block         handle_block ] Got block #10076 from network                       application.cp
p:356
2015-09-01T11:24:30 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-09-01T11:24:35 th_a:invoke handle_block         handle_block ] Got block #10077 from network                       application.cp
p:356
2015-09-01T11:24:35 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
item->num < _head->num + MAX_BLOCK_REORDERING:
    {}
    th_a  fork_database.cpp:70 _push_block

    {"new_block":{"previous":"0000275c210957d57af8b062f9b181a388f4f156","timestamp":"2015-09-01T11:24:35","witness":"1.6.14","transact
ion_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"2026a8ae6f9d70675e0db87eaffbfd59c5c35
689f53d7b1fee8c0ebe5e1915fea9772b5147c0b11ca94017840c243facdf75d3cdacc3cab2cc0e8111fa2e3f3d43","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:378
...
...
...
2015-09-01T11:26:26 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
item->num > std::max<int64_t>( 0, int64_t(_head->num) - (_max_size) ): attempting to push a block that is too old
    {"item->num":10046,"head":10076,"max_size":10}
    th_a  fork_database.cpp:69 _push_block

    {"new_block":{"previous":"0000273d2323d0bf7fee2b07a8dc945512fac8d8","timestamp":"2015-09-01T11:22:00","witness":"1.6.1","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"200d1bf2caeeba16ebc2cb5a505e2ea2c8df4d6c1193fd74e9e961acf85b54826c782e18f8e233b5edb589b02e304cdeffd6fda2091fe07967d5db26baf412ccad","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:378
```

**Comment by bytemaster** [comment]
I just increased the MAX_BLOCK_REORDERING to 1024 which should give the network much more time to recover from a timeout.

**Comment by abitmore** [comment]
Looks like it's not fixed completely. If `MAX_BLOCK_REORDERING` was 32, what is this `"max_size":10` (the assert at fork_database.cpp:69)?

**Comment by abitmore** [comment]
Still have this issue with tag `test2b`:

```
2015-09-02T02:12:15 th_a:invoke handle_block         handle_block ] Got block #20187 from network                       application.cpp:356
2015-09-02T02:12:15 th_a:invoke handle_block   on_objects_changed ] ids: ["2.8.20187","1.6.64","1.13.72","2.1.0"]
        api.cpp:867
2015-09-02T02:12:15 th_a:invoke handle_block   on_objects_changed ] ids: ["2.8.20187","1.6.64","1.13.72","2.1.0"]
        api.cpp:867
2015-09-02T02:12:25 th_a:invoke handle_block         handle_block ] Got block #20189 from network                       application.cpp:356
2015-09-02T02:12:25 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
2015-09-02T02:12:30 th_a:invoke handle_block         handle_block ] Got block #20190 from network                       application.cpp:356
2015-09-02T02:12:30 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.
        fork_database.cpp:57
...
...
...
2015-09-02T02:20:22 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
item->num > std::max<int64_t>( 0, int64_t(_head->num) - (_max_size) ): attempting to push a block that is too old
    {"item->num":20188,"head":20278,"max_size":10}
    th_a  fork_database.cpp:69 _push_block

    {"new_block":{"previous":"00004edb437dfbc741e818a8a9754a653da79111","timestamp":"2015-09-02T02:12:20","witness":"1.6.3706","transaction_merkle_root":"0000000000000000000000000000000000000000","extensions":[],"witness_signature":"1f0f14d642abd9b0d26e0d2e5960ad5b4e1d340282317fd1948f942e1f857804393bbeca54636b9f3ad68e44f9843648aa403fb5dc13024649d259558db3c119f1","transactions":[]}}
    th_a  db_block.cpp:176 _push_block                  application.cpp:378
```

**Comment by bytemaster** [comment]
max_size represents the maximum number of blocks in the past that may be linked to.
MAX_BLOCK_REORDERING represents the number of unlinked blocks in the future.

**Comment by theoreticalbts** [comment]
May be fixed by fix for #299.  Feel free to re-open if it re-surfaces.
